src/main/java/com/videoplaza/dataflow/pubsub/PubsubSourceTask.java

```diff
@@ -0,0 +1,318 @@
+package com.videoplaza.dataflow.pubsub;
+
+import com.google.api.core.ApiService;
+import com.google.cloud.pubsub.v1.AckReplyConsumer;
+import com.google.cloud.pubsub.v1.Subscriber;
+import com.google.pubsub.v1.PubsubMessage;
+import org.apache.kafka.connect.source.SourceRecord;
+import org.apache.kafka.connect.source.SourceTask;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static com.videoplaza.dataflow.pubsub.TimeUtils.msSince;
+import static com.videoplaza.dataflow.pubsub.TimeUtils.msTo;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * TODO report metrics
+ */
+public class PubsubSourceTask extends SourceTask {
+
+   private static final AtomicInteger TASK_COUNT = new AtomicInteger();
+
+   private final int id = TASK_COUNT.incrementAndGet();
+   private final Logger log = LoggerFactory.getLogger(PubsubSourceTask.class.getName() + "-" + id);
+
+   private final TaskMetrics metrics = new TaskMetrics();
+   private final Map<String, MessageInFlight> messages = new ConcurrentHashMap<>();
+   private final Set<PubsubMessage> toBePolled = ConcurrentHashMap.newKeySet();
+
+   private final AtomicBoolean stopping = new AtomicBoolean();
+   private final Lock pollCommitLock = new ReentrantLock();
+   private final Lock receiveLock = new ReentrantLock();
+   private final Condition recordsReceived = receiveLock.newCondition();
+
+   private volatile Subscriber subscriber;
+   private volatile PubsubSourceConnectorConfig config;
+   private volatile PubsubMessageConverter converter;
+   private volatile Sleeper sleeper;
+
+   @Override public String version() {
+      return Version.getVersion();
+   }
+
+   @Override public void start(Map<String, String> props) {
+      configure(props);
+      configure(new SimpleSleeper());
+      subscribe(newSubscriber());
+      log.info("Started");
+   }
+
+   PubsubSourceTask configure(Sleeper sleeper) {
+      this.sleeper = sleeper;
+      return this;
+   }
+
+   Subscriber newSubscriber() {
+
+      Subscriber subscriber = Subscriber.newBuilder(config.getProjectSubscription(), this::onPubsubMessageReceived)
+          .setFlowControlSettings(config.getFlowControlSettings())
+          .build();
+
+      subscriber.addListener(new Subscriber.Listener() {
```

Comment:
This could be broken out to it's own class to reduce this big-ish class and improve readability?

Reply:
Sure. Its still gonna be inner class though since it is pretty internal.